src/models/s2s.h

```diff
@@ -194,29 +196,56 @@ class DecoderS2S : public DecoderBase {
              && opt<std::string>("original-type") == "nematus")  //
         ("skip", opt<bool>("skip"));
 
+    ABORT_IF(opt<bool>("dec-attention-bilinear-lookup"),
+             "--dec-attention-bilinear-lookup is disabled due to a bug");
+
     size_t decoderLayers = opt<size_t>("dec-depth");
     size_t decoderBaseDepth = opt<size_t>("dec-cell-base-depth");
     size_t decoderHighDepth = opt<size_t>("dec-cell-high-depth");
+    int decoderAttentionHops = opt<int>("dec-attention-hops");
+    ABORT_IF(decoderAttentionHops >= decoderBaseDepth,
+             "--dec-attention-hops must be lower than dec-cell-base-depth");
 
     // setting up conditional (transitional) cell
     auto baseCell = rnn::stacked_cell(graph);
-    for(size_t i = 1; i <= decoderBaseDepth; ++i) {
-      bool transition = (i > 2);
+
+    lastAttentionCellIds_ = std::vector<size_t>(state->getEncoderStates().size(), -1);
+    numBaseCells_ = 0;
+    for(int i = 1; i <= decoderBaseDepth; ++i) {
+      bool transition = (i > 1+decoderAttentionHops);
       auto paramPrefix = prefix_ + "_cell" + std::to_string(i);
       baseCell.push_back(rnn::cell(graph)         //
                          ("prefix", paramPrefix)  //
                          ("final", i > 1)         //
                          ("transition", transition));
-      if(i == 1) {
-        for(size_t k = 0; k < state->getEncoderStates().size(); ++k) {
+
+      ++numBaseCells_;
+      //LOG(info, "Decoder, create cell: i:{} paramPrefix:{} transition:{} final:{}", i, paramPrefix, transition, i > 1);
+      if(i <= decoderAttentionHops) {
+        for(int k = 0; k < state->getEncoderStates().size(); ++k) {
           auto attPrefix = prefix_;
           if(state->getEncoderStates().size() > 1)
             attPrefix += "_att" + std::to_string(k + 1);
+          if (i > 1) {
+            attPrefix += "_hop" + std::to_string(i);
+          }
+          //LOG(info, "Decoder, create attention: {}", attPrefix);
 
           auto encState = state->getEncoderStates()[k];
 
-          baseCell.push_back(
-              rnn::attention(graph)("prefix", attPrefix).set_state(encState));
+         
+          baseCell.push_back(rnn::attention(graph)("prefix", attPrefix)
+                                   ("attentionHeads", opt<int>("dec-attention-heads"))
+                                   ("attentionLookupDim", opt<int>("dec-attention-lookup-dim"))
+                                   ("attentionProjectionDim", opt<int>("dec-attention-projection-dim"))
+                                   ("attentionIndependentHeads", opt<bool>("dec-attention-independent-heads"))
+                                   ("attentionBilinearLookup", opt<bool>("dec-attention-bilinear-lookup"))
+                                   ("attentionProjectionLayerNorm", opt<bool>("dec-attention-projection-layernorm"))
+                                   ("attentionProjectionTanH", opt<bool>("dec-attention-projection-tanh"))
```

Comment:
I think we instead use a string to denote the non-linearity, e.g. cf. `transformer-ffn-activation`. Please follow the same pattern.

Reply:
I had only implemented a tanh here.

Reply:
Can you move the `activationByName()` function [https://github.com/marian-nmt/marian-dev/blob/38722782de7b3a247e53e5218a9b259f3d740f8c/src/models/transformer.h#L344] to a shared place and then implement the same parameterization pattern as in Transformer based on it? Example use:

    auto actFn = activationByName(opt<std::string>("transformer-ffn-activation"));